### PR TITLE
Make users able to clear workspace leases

### PR DESCRIPTION
**Pull request by res0nance**
Add ability to release workspace leases. This helps when a hardkill happens and the user would like to reset their workspace leases so they do not see `@2` appended to the end of their workspace

### Proposed changelog entries

* Entry 1: Add button to clear singe workspace leases in the computer page
* ...

<details close>
<summary>Screenshots</summary>
No leases:

image

Single lease:

image
</details>

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+  <l:layout>
+		<st:include page="sidepanel.jelly" />
```
Comment (timja) [inline core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly]: Indenting looks extreme here?

**Review by timja — DISMISSED** [review]

File: core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly (hunk 1 above)
Comment (res0nance) [reply core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly]: Got ambushed by tabs, re-indenting this file

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Needs a lot more documentation at a minimum (and even then I'm not sure this is useful enough in general).

When would someone do this? How will this impact running builds? Can't Jenkins figure this out itself and only offer this when it matters?

**Comment by res0nance** [comment]
> When would someone do this? How will this impact running builds? Can't Jenkins figure this out itself and only offer this when it matters?

They would do this when scripts cannot withstand the suffix addition. This addition is sometimes caused by doing `/kill` on builds.

I can make it only appear when there are used workspaces but I can't release it automagically( at least I don't know of a way ).

**Comment by daniel-beck** [comment]
> They would do this when

The point of the questions is to illustrate what users should be able to figure out.

If you're writing this UI only for yourself, write a plugin instead. If this is for all Jenkins users, it needs to make sense to all Jenkins users.

**Comment by res0nance** [comment]
> If you're writing this UI only for yourself, write a plugin instead. If this is for all Jenkins users, it needs to make sense to all Jenkins users.

That makes sense, perhaps this should be more like lockable resources which shows the build that holds it. Probably a lot more useful

OTOH: Perhaps this just needs some special background checking in case of a hardkill to ensure that these workspaces are cleaned up.

I'm open to ideas and suggestions. My users seem to experience this issue every now and then and I'm sure many others do too. So I'm looking for an idea of how to improve things for all users.

File: core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+    <l:layout>
+        <st:include page="sidepanel.jelly" />
+        <l:main-panel>
+            <form method="post" action="doReleaseWorkspaces">
+            ${%Release workspaces from }${it.name}
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/Computer/releaseWorkspaces.jelly]: Should be a parameterized localized message.

**Comment by daniel-beck** [comment]
@res0nance I agree this is a problem that needs addressing. Obviously the best approach would be obviating the need for hard kills in pipeline, but I doubt we'll get that any time soon.

My request here is two-fold:

* Can we try to decide more intelligently whether to offer this, and where? For example, can we detect leases with no owner, leases with an owning build that is finished, aborted, etc? It's been years since I looked at the API, I don't know whether this even makes sense.
* Inline docs and explanations (and I don't think we've exposed "workspace leases" to the UI yet) that allow at least a large subset of Jenkins admins to understand what this is for.

> That makes sense, perhaps this should be more like lockable resources which shows the build that holds it. Probably a lot more useful

That would be a big improvement.

Note that one thing that's a potential concern here is runaway processes still working on the workspace. I.e. a long running shell script that hasn't gotten the signals, potentially deliberately, when the build was killed, and may continue writing to the workspace.

Another option might be to integrate this into the workspace browser, similar to wiping workspaces, or as "Wipe & release"?

Haven't tried any of this, just some ideas.

**Comment by res0nance** [comment]
@daniel-beck I've worked on this, pipeline uses a blank object but freestyle seems to use the build as its context object. To get its actual run there might need to be downstream changes in pipeline to make this all look good.

**Comment by varyvol** [comment]
@daniel-beck it seems @res0nance added some changes that you have not discussed. Could you maybe have a look so that this moves forward?

**Comment by varyvol** [comment]
Given there's been no activity for some time, we are proposing this for close. Please let us know in case you are actually interested in continue working on this. Thanks!

**Comment by res0nance** [comment]
I think this needs a little rethink, jenkins should figure this out automagically. So I will close this. Thanks

**Pull request closed without merging**